### ISSUE TITLE
RFC: Synchronous-promise versioning after rewrite

### ISSUE THREAD

**Issue by fluffynuts** [ctavan/jupp#11]
Hi

Please could you leave your thoughts on (https://github.com/fluffynuts/synchronous-promise/issues/7) as I have a fix for that issue which involved a rewrite of the code, maintaining the original tests, and I'd like to minimise the fallout, if any (:

**Comment by fluffynuts** [comment]
2.0.0 is released. Please update at your convenience.

**Comment by ctavan** [comment]
Thanks @fluffynuts, I've pulled your changes in, they are released as v0.28.0.

Seems like it has become a bit slower though.

Before:

```
$ babel-node benchmark/index.js
schema.validate x 4,799 ops/sec ±4.48% (54 runs sampled)
schema.validateSync x 2,323 ops/sec ±3.19% (75 runs sampled)
schema.cast x 6,243 ops/sec ±2.88% (81 runs sampled)
✨  Done in 27.72s.
```

After:

```
schema.validate x 4,648 ops/sec ±3.83% (60 runs sampled)
schema.validateSync x 1,513 ops/sec ±3.65% (69 runs sampled)
schema.cast x 6,144 ops/sec ±4.42% (74 runs sampled)
✨  Done in 29.16s.
```

Since I'm currently not so much concerned about performance I don't mind though.

**Comment by fluffynuts** [comment]
If you have a lot of promise-stuff going on, it could well end up a little slower -- though I'm not sure from the output how many actual invocations are happening.

The reason is (probably) quite simple: 
Previously, `synchronous-promise` basically returned itself on every `.then()` and `.catch()`, storing an array of functions to call in order when resolving (with a little faerie-dust to ensure that promise-like results were handled properly even if they were _proper_ (read: async) promises). 
This works well when starting a test-driven approach where one is trying to solve the regular use-cases (handling `.then()`, `.catch()` and the optional failure function which can be supplied to `then()`). It starts to show weakness when creating diverging paths from a promise, eg, using 1.0.x:
```
var promise = SynchronousPromise.resolve("moo");
promise.then(result => "cow").then(result => console.log(result)); // prints "cow"
promise.then(result => console.log(result)); // also prints "cow", but should print "moo"
```
Which showed up in reported issue https://github.com/fluffynuts/synchronous-promise/issues/7. It may not be a super-common use-case to re-use a promise, but it's obviously an expected behavior. Hence the desire to rewrite, after much thinking about how I could cheat a little got me nowhere (:

The new code creates new `SynchronousPromise`s from each `.then()` and `.catch()` call so that divergent paths can be properly handled. So if you're creating hundreds or thousands of them, I should expect that they would be _slightly_ slower than before, since we have whole object creation instead of just handler storage / calling happening.
